Prototype CloudFormation template

jeancochrane opened the pull request:
# Overview

Ports the Terraform stack to a CloudFormation template that can be used to deploy AWS resources.

Closes https://github.com/azavea/raster-vision-aws/issues/16.

## Demo

Parameters can now be configured via the Cloudformation UI:

<img width="1552" alt="screen shot 2018-12-18 at 12 49 48 pm" src="https://user-images.githubusercontent.com/14170650/50172613-779a7400-02c3-11e9-8662-555e2126a1e6.png">

## Notes

- This PR seeks to make as few changes as possible to the resource design -- instead, it simply seeks to port existing functionality to CloudFormation. Any changes from the existing infrastructure will be highlighted in comments in the code.

# Testing instructions

- Log into the R&D AWS console
- Navigate to `CloudFormation > Create Stack`
- In the `Choose a template field`, select `Upload a template to Amazon S3` and upload the template in `cloudformation/template.yml`
- Where available, use the default parameters. Specify the following required parameters:
    - `Stack Name`: `TestingCloudFormation`
    - `VPC`: `vpc-074e8ae6256f7c1ca` (the first one on the dropdown)
    - `Subnets`: `subnet-033d2cbe8268c3067` (the first one on the dropdown)
    - `SSH Key Name`: `raster-vision` (doesn't matter unless you want to shell into the instance)
- Accept all default options
- Accept `I acknowledge that AWS CloudFormation might create IAM resources with custom names` on the last screen
- Create the stack and confirm that it builds correctly
- Delete the stack and confirm that everything cleans up correctly

Commit:
Sketch out CloudFormation template

Take a first pass at converting Terraform infrastructure to a
CloudFormation template. Note that the temlpate hasn't yet been
tested, and likely contains errors as of this commit.

Commit:
Add template metadata and fix syntax errors

Use template metadata to specify the order and grouping of parameters
in the AWS CloudFormation console. In addition, fix syntax errors in
the template raised by testing.

jeancochrane commented on cloudformation/template.yml:
```diff
@@ -0,0 +1,339 @@
+    Type: 'AWS::EC2::SecurityGroup'
+    Properties:
+      VpcId: !Ref VPC
+      GroupDescription: !Join ['', ['Security group for ', !Ref Project, ' (', !Ref Environment, ')']]
```
`GroupDescription` is a [required attribute](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-security-group.html#cfn-ec2-securitygroup-groupdescription) in CloudFormation.

hectcastro replied:
Yeah, I think that Terraform defaults it to something like `Managed by Terraform`.

jeancochrane commented on cloudformation/template.yml:
```diff
@@ -0,0 +1,339 @@
+
+  Environment:
+    Type: String
+    Default: Testing
```
I don't necessarily like this default (or the others that are set in this file), but I adjusted them in order to prevent clobbering existing resources which already use the `Staging` namespace.

hectcastro replied:
We've used `Unknown` for this value in Terraform modules. Not great, but at least it communicates that you should change it.

jeancochrane replied:
I like that, much clearer.

hectcastro commented on cloudformation/template.yml:
```diff
@@ -0,0 +1,339 @@
+    Description: Environment of the project being deployed (e.g. Staging, Development, Production)
+
+  Region:
+    Type: String
```
I think that we can use `AWS::EC2::AvailabilityZone::Name` here.

jeancochrane replied:
I did a quick scan of the template to make sure that would work, and it appears that we don't use this region anywhere (the Terraform stack only uses it in `config.tf` to set up the provider). I'll just go ahead and remove it, since the `Subnets` control the availability zone of the actual resources.

hectcastro commented on cloudformation/template.yml:
```diff
@@ -0,0 +1,339 @@
+    Description: AWS Region in which to deploy the Batch job
+
+  AMI:
+    Type: String
```
Same here, but `AWS::EC2::Image::Id`.

hectcastro commented on cloudformation/template.yml:
```diff
@@ -0,0 +1,339 @@
+    Description: Amazon Machine Image to use for the compute environment
+
+  KeyName:
+    Type: 'AWS::EC2::KeyPair::KeyName'
```
I think that we can safely remove the quoting here.

jeancochrane replied:
Roger! I went for it because I was worried that `:` always acted as a special character in YAML, but [that appears not to be true](https://stackoverflow.com/a/33149072).

hectcastro commented on cloudformation/template.yml:
```diff
@@ -0,0 +1,339 @@
+
+  SpotFleetBidPercentage:
+    Type: Number
+    Default: 100
```
Last time I ran through the Batch console, it defaults this to 40.

jeancochrane replied:
Sounds reasonable to me. This is another case of the default that was set in `raster-vision-aws`: https://github.com/azavea/raster-vision-aws/blob/b7e29cb79f4d99a67f134c6dde9177bf37dbae60/terraform/variables.tf#L17-L19

hectcastro commented on cloudformation/template.yml:
```diff
@@ -0,0 +1,339 @@
+
+  InstanceTypes:
+    Type: List<String>
+    Default: p2.xlarge
```
It might be useful to add an `AllowedValues` attribute here for only the GPU instance types we want to target.

```yaml
AllowedValues:
  - ...
  - ...
```

jeancochrane replied:
Nice.

jeancochrane replied:
Having some trouble getting `AllowedValues` to work properly with the `List<String>` type here. When `AllowedValues` is specified, CloudFormation always throws the following template syntax error, no matter the default value passed in:

> Template validation error: Each value of parameter 'InstanceTypes' must be one of AllowedValues

Here's the relevant block:

```yaml
  InstanceTypes:
    Type: List<String>
    Default: p2.xlarge
    AllowedValues:
      - p2.xlarge
      - p2.8xlarge
      - p2.16xlarge
```

No docs on this online as far as I can tell, either -- every example I could find uses `AllowedValues` for a `String` or `Number` type, which makes me suspect that it's not supported for `List`s. Are you comfortable punting on this until we get a sense of how useful it would be?

hectcastro replied:
This was the snippet I was referencing in standup:

https://github.com/raster-foundry/prediction-api/blob/a7a69abb29ce9f9c33e27ec1e9dbc8a6c732e047/deployment/cloudformation/hot-prediction-endpoint-template.yaml#L17-L27

I think the point you made still stands though. This example is not of type list.

jeancochrane replied:
Good to know! I'll post a question on the AWS Developer Forums (once my account has permission to post new topics) to see if I can get an answer about whether it's possible, but for now I'll remove `AllowedValues`.

hectcastro commented on cloudformation/template.yml:
```diff
@@ -0,0 +1,339 @@
+      Tags:
+        -
+          Key: Name
+          Value: !Join ['', ['sgRasterVisionContainerInstance', !Ref Environment]]
```
I think we are OK omitting the environment from the security group name here.

hectcastro commented on cloudformation/template.yml:
```diff
@@ -0,0 +1,339 @@
+Description: A CloudFormation template for deploying Raster Vision Batch jobs to AWS.
+
+Metadata:
+  'AWS::CloudFormation::Interface':
```
Interesting. I was not aware that this existed. This should help out the unfamiliar user quite a bit.

hectcastro commented on cloudformation/template.yml:
```diff
@@ -0,0 +1,339 @@
+    Default: 55000
+    Description: The hard limit (in MB) of memory to present to the container
+
+  BatchServiceRolePolicyARN:
```
Given the intended audience, I could see there being value in omitting these policy ARNs from the parameters and just hardcoding them into the roles we create below.

hectcastro reviewed (commented):
Good start. The YAML looks pretty clean.

Commit:
Remove unnecessary resources

A few resources and defaults were unnecessary for this template.
Remove them, alongside a general style cleanup.

Commit:
Remove unnecessary quotation marks from cloudformation template

Previous commits introduced superstitious quotation marks around
strings that used YAML special characters in the cloudformation
template file. Remove those quotations marks in cases where YAML
will properly interpret the characters as strings.

Commit:
Remove AllowedValues from InstanceType parameter

It is currently unclear whether CloudFormation supports the
AllowedValues directive for input parameters of the List type.
For now, remove it to avoid esoteric template validation errors.

hectcastro reviewed (approved):
👍 I think we are in a good position to ask R&D to go through the workflow of standing up a new Batch environment to see how they think this streamlines things.

Might be valuable to expand on your PR description a bit in a `README` for this repository so that people know how to apply this template via the CloudFormation console.

jeancochrane committed:
Add CloudFormation instructions to the README

Add a short section to the README documenting the basic steps a user would need to take to deploy infrastructure using the CloudFormation template.

Commit:
Set the default ECR_IMAGE variable to match the CloudFormation stack

Make sure that ECR_IMAGE points to the name of the built image
supplied in the CloudFormation stack. Otherwise, Batch will not
pull the correct resource on startup, unless the user is savvy enough
to adjust this variable.

jeancochrane commented:
Awesome! Added a short section to the README documenting the CloudFormation setup. Will pull this in now, and send off an email to Rob and Lewis first thing tomorrow soliciting feedback.
